compiler/src/iree/compiler/InputConversion/MHLO/test/convert_collective_ops.mlir

```diff
@@ -0,0 +1,167 @@
+// RUN: iree-opt --split-input-file --iree-mhlo-to-linalg-on-tensors --canonicalize -cse %s | FileCheck %s
+
+// CHECK-LABEL: @replica_id
+func.func @replica_id() -> tensor<ui32> {
+  // CHECK-DAG: [[CHANNEL:%.+]] = flow.channel.default : !flow.channel
+  // CHECK-DAG: [[RANK:%.+]] = flow.channel.rank [[CHANNEL]] : index
+  // CHECK-DAG: [[CAST:%.+]] = arith.index_castui [[RANK]] : index to i32
+  // CHECK-DAG: [[TENSOR:%.+]] = tensor.from_elements [[CAST]] : tensor<i32>
```

Comment:
This is a great catch! I wrote a test program, and I found that `linalg-detensorize` didn't optimize the case by default. When I use the pass's sub-option, `aggressive-mode=true`, it detensorizes correctly.

Here is the input and the detensorized output:
```mlir
// -----// IR Dump After FusionOfTensorOps (iree-flow-fusion-of-tensor-ops) //----- //
func.func @replica_id_dynamic_slice(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
  %c2_i32 = arith.constant 2 : i32
  %c512_i32 = arith.constant 512 : i32
  %c0 = arith.constant 0 : index
  %c512 = arith.constant 512 : index
  %0 = hal.tensor.import %arg0 : !hal.buffer_view -> tensor<1024xf32>
  %channel_default = flow.channel.default : !flow.channel
  %channel_rank = flow.channel.rank %channel_default : index
  %1 = arith.index_castui %channel_rank : index to i32
  %from_elements = tensor.from_elements %1 : tensor<i32>
  %2 = tensor.empty() : tensor<i32>
  %3 = linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%from_elements : tensor<i32>) outs(%2 : tensor<i32>) {
  ^bb0(%in: i32, %out: i32):
    %8 = arith.remui %in, %c2_i32 : i32
    %9 = arith.muli %8, %c512_i32 : i32
    linalg.yield %9 : i32
  } -> tensor<i32>
  %extracted = tensor.extract %3[] : tensor<i32>
  %4 = arith.index_castui %extracted : i32 to index
  %5 = arith.maxsi %4, %c0 : index
  %6 = arith.minsi %5, %c512 : index
  %extracted_slice = tensor.extract_slice %0[%6] [512] [1] : tensor<1024xf32> to tensor<512xf32>
  %7 = hal.tensor.export %extracted_slice : tensor<512xf32> -> !hal.buffer_view
  return %7 : !hal.buffer_view
}

// -----// IR Dump After LinalgDetensorize (linalg-detensorize) //----- //
func.func @replica_id_dynamic_slice(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
  %c2_i32 = arith.constant 2 : i32
  %c512_i32 = arith.constant 512 : i32
  %c0 = arith.constant 0 : index
  %c512 = arith.constant 512 : index
  %0 = hal.tensor.import %arg0 : !hal.buffer_view -> tensor<1024xf32>
  %channel_default = flow.channel.default : !flow.channel
  %channel_rank = flow.channel.rank %channel_default : index
  %1 = arith.index_castui %channel_rank : index to i32
  %2 = arith.remui %1, %c2_i32 : i32
  %3 = arith.muli %2, %c512_i32 : i32
  %4 = arith.index_castui %3 : i32 to index
  %5 = arith.maxsi %4, %c0 : index
  %6 = arith.minsi %5, %c512 : index
  %extracted_slice = tensor.extract_slice %0[%6] [512] [1] : tensor<1024xf32> to tensor<512xf32>
  %7 = hal.tensor.export %extracted_slice : tensor<512xf32> -> !hal.buffer_view
  return %7 : !hal.buffer_view
}
```

Reply:
@MaheshRavishankar do you happen to know why the default detensorizer fails? Also, how can I enable the sub-option in the command line? Thanks!

Reply:
awesome, thanks for trying this! would indeed be really good to get it turned on (has been needed for a very long time!), or at least ensure you're doing your testing with it on until it can be on by default so you don't end up measuring the slow stuff :)